zk_prover/src/circuits/utils.rs

```diff
@@ -33,11 +33,32 @@ use std::{
     path::PathBuf,
 };
 
-/// Generate setup parameters for a circuit of size `k` where 2^k is the number of rows in the circuit.
-/// First checks if the trusted setup parameters are already generated and saved in the `ptau` folder with the name `hermez-raw-k`, if so, it loads them.
-/// If the trusted setup are not found in the `ptau` folder, an unsafe trusted setup is generated and saved there.
-pub fn generate_setup_params(k: u32) -> ParamsKZG<Bn256> {
-    let ptau_path = format!("ptau/hermez-raw-{}", k);
+/// Generate setup artifacts for a circuit of size `k`, where 2^k represents the number of rows in the circuit.
+/// This function first checks if the trusted setup parameters are already generated and saved in the `ptau` folder with the name `hermez-raw-k`.
+///
+/// If the trusted setup parameters are not found, the function performs an unsafe trusted setup to generate the necessary parameters
+/// If the provided `k` value is larger than the `k` value of the loaded parameters, an error is returned, as the provided `k` is too large.
+/// Otherwise, if the `k` value is smaller than the `k` value of the loaded parameters, the parameters are downsized to fit the requested `k`.
+pub fn generate_setup_artifacts<C: Circuit<Fp> + CircuitExt<Fp>>(
+    k: u32,
+    params_path: Option<&str>,
+    circuit: C,
+) -> Result<
+    (
+        ParamsKZG<Bn256>,
+        ProvingKey<G1Affine>,
+        VerifyingKey<G1Affine>,
+    ),
+    &'static str,
+> {
+    let mut ptau_path = format!("ptau/hermez-raw-{}", k);
```

Comment:
Can we remove this line and any reference to `ptau_path`? I think that the `generate_setup_artifacts` function should only consider two cases: 

- Case in which no `params_path` is provided as input. In that case the trusted setup is generated randomly.
- Case in which `params_path` is provided as input. In that case the trusted setup is generated starting from the ptau file from `params_path`

Reply:
If you think that this modification makes sense, please also update the description on how to generate the trusted setup in the README

Reply:
Totally makes sense. I was wrong in checking the `ptau` file in that function.

Reply:
Fixed at [42b1dba](https://github.com/summa-dev/summa-solvency/pull/117/commits/42b1dbad462239eb315390299bf33e84148aa294)